### PR TITLE
Implement private protected access

### PR DESCRIPTION
**Pull request by gafter**
This is the main implementation of the language feature `private protected` (#1384)

This PR is for review only (it will not be integrated into future). Once this is reviewed, future PRs for this feature will be incremental to this, as PRs into this feature branch.

@dotnet/roslyn-compiler Please review

### CODE REVIEW

File: docs/specs/PortablePdb-Metadata.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#Portable PDB v1.0: Format Specification
+# Portable PDB v1.0: Format Specification
```
Comment (AlekseyTs) [inline docs/specs/PortablePdb-Metadata.md]: This change feels unrelated to the feature. Revert?

File: src/Compilers/CSharp/Portable/Compiler/ClsComplianceChecker.cs [hunk 2]
```diff
@@ -1012,7 +1012,10 @@ private bool IsCompliantType(NamedTypeSymbol type, NamedTypeSymbol context)
         private static bool IsInaccessibleBecauseOfConstruction(NamedTypeSymbol type, NamedTypeSymbol context)
         {
             // NOTE: Dev11 (incorrectly) only checks whether "type" is protected - it ignores container accessibility.
-            bool sawProtected = type.DeclaredAccessibility == Accessibility.Protected || type.DeclaredAccessibility == Accessibility.ProtectedOrInternal;
+            bool sawProtected =
+                type.DeclaredAccessibility == Accessibility.Protected ||
+                type.DeclaredAccessibility == Accessibility.ProtectedAndInternal ||
+                type.DeclaredAccessibility == Accessibility.ProtectedOrInternal;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/ClsComplianceChecker.cs]: Would it make sense to cache ```type.DeclaredAccessibility``` in a local and reuse?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/ClsComplianceChecker.cs]: It might be better to extract this into a method taking Accessibility, there is another use site below.

File: src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs [hunk 3]
```diff
@@ -215,11 +217,19 @@ public static DeclarationModifiers ToDeclarationModifiers(this SyntaxTokenList m
...
             {
-                result &= ~DeclarationModifiers.AccessibilityMask;
-                result |= DeclarationModifiers.ProtectedInternal;
+                case DeclarationModifiers.Protected | DeclarationModifiers.Internal:
+                    // the two keywords "protected" and "internal" together are treated as one modifier.
+                    result &= ~DeclarationModifiers.AccessibilityMask;
+                    result |= DeclarationModifiers.ProtectedInternal;
+                    break;
+
+                case DeclarationModifiers.Private | DeclarationModifiers.Protected:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs]: Do we need to handle 
```
case DeclarationModifiers.Private | DeclarationModifiers.Protected | DeclarationModifiers.Internal:
```
?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 4]
```diff
@@ -1810,7 +1818,7 @@ private void CheckForProtectedInStaticClass(DiagnosticBag diagnostics)
                         continue;
                     }
 
-                    if (member.DeclaredAccessibility == Accessibility.Protected || member.DeclaredAccessibility == Accessibility.ProtectedOrInternal)
+                    if (member.DeclaredAccessibility == Accessibility.Protected || member.DeclaredAccessibility == Accessibility.ProtectedOrInternal || member.DeclaredAccessibility == Accessibility.ProtectedAndInternal)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Here is the third place for this condition.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 5]
```diff
@@ -894,6 +894,7 @@ private static bool AddHidingAbstractDiagnostic(Symbol hidingMember, Location hi
             {
                 case Accessibility.Internal:
                 case Accessibility.Private:
+                case Accessibility.ProtectedAndInternal:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: It is not obvious why this shouldn't be handled the same way as other protected.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 6]
```diff
@@ -644,6 +656,23 @@ private static bool IsAsRestrictive(NamedTypeSymbol s1, Symbol sym2, ref HashSet
...
+                                    // If s2 is in the same assembly as s1, then s2's private protected is more restrictive
+                                    // than s1's internal protected.
+                                    if (s2.ContainingAssembly.HasInternalAccessTo(s1.ContainingAssembly))
+                                    {
+                                        return true;
+                                    }
+
+                                    // if s2 is private protected, and it's parent is a subclass (or the same as) s1's parent
+                                    // then this is at least as restrictive as s1's internal protected
+                                    if (parent1.IsAccessibleViaInheritance(s2.ContainingType, ref useSiteDiagnostics))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: Can we simply jump to ```case Accessibility.Protected:``` for this check?

File: src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs]: No, we don't want to give any particular meaning to that combination; we want to let the caller report it as an error.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: The principal difference between these two categories of access levels is that these ones do not cross assembly boundaries, and the other ones do. Specifically, a `private protected` member that hides a virtual member does not prevent the hidden member from being implemented (in another assembly). But a `protected` member that hides a virtual member *does* prevent it from being implemented.

Having said that, it isn't clear to me what language specification rule justifies the diagnostics produced here. I'll file an issue.

File: src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs]: Do we have a unit-test covering this situation?

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb [hunk 7]
```diff
@@ -187,7 +188,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                 End If
             Else
                 ' Nested types (including types in modules) can be any accessibility, and the default is public
-                If (modifiers And DeclarationModifiers.Private) <> 0 Then
+                If (modifiers And (DeclarationModifiers.Private Or DeclarationModifiers.Protected)) =
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: Should we handle the case of all three bits set?

**Comment by AlekseyTs** [comment]
There is a number of TODO comments that, which should be switched to PROTOTYPE.

File: src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs]: This is an "initial prototype" in the terminology of https://gist.github.com/gafter/c8825a0c889b6a5d6999 , and you can see the .work.md file included in this PR describes the testing scenarios remaining (which is basically each of the specified scenarios). The tests you're asking about would not be classified as among those that are part of the initial prototype.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb (hunk 7 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: No, I don't think so. The caller already handles that by reporting extra access modifiers.

**Comment by gafter** [comment]
Actually, the two TODOs are about missing access levels in `EnvDTE`. Not something we can fix in the compiler.

**Comment by jaredpar** [comment]
> Actually, the two TODOs are about missing access levels in EnvDTE. Not something we can fix in the compiler.

It sounds though like this is still work that needs to happen.  Hence it needs to be tracked in a way that will be resolved when / if the branch is merged into future.  We should be filing bugs / using PROTOTYPE to track this.

**Comment by gafter** [comment]
Where would I file a bug against EnvDTE?

**Comment by jaredpar** [comment]
Inside the VSO database.  Seems premature to do htat though until we commit to this feature which is why I agreed with the PROTOTYPE solution.

**Comment by gafter** [comment]
Given that this is a valid CLR access level, it would appear to be a bug that it is not supported in EnvDTE.vsCMAccess, whether C# exposes it or not.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler All code review comments answered or responded to by changes. Please re-review.

**Comment by gafter** [comment]
@dotnet-bot retest prtest/win/dbg/eta please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_dbg_eta/5337/
// Retest reason:

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Can I please have a second sign-off on this? This is for the initial code in the feature branch, not for integration into a product branch.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 8]
```diff
@@ -560,6 +560,18 @@ private static bool IsAsRestrictive(NamedTypeSymbol s1, Symbol sym2, ref HashSet
                             break;
                         }
 
+                    case Accessibility.ProtectedAndInternal:
+                        // Since s1 is private protected, s2 must pass the test for being both more restrictive than internal and more restrictive than protected.
+                        // We first do the "internal" test (copied from above), then if it passes we continue with the "protected" test.
+
+                        if ((acc2 == Accessibility.Private || acc2 == Accessibility.Internal || acc2 == Accessibility.ProtectedAndInternal) && s2.ContainingAssembly.HasInternalAccessTo(s1.ContainingAssembly))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: Would it be worth it to factor two helper methods out instead? One for `IsAsRestrictiveAsInternal` and one for `IsAsRestrictiveAsProtected`.
Then the code will be more self-explanatory.

**Comment by jcouv** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs (hunk 8 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: I think it would be more confusing due to the number of distinct locals that are used and would be passed as parameters. Also, this is inner-loop stuff, so I hesitate to extract method.

**Comment by gafter** [comment]
Closing as review complete. The `features/privateProtected` branch has now been reviewed.
